### PR TITLE
Feature/distinguish smartlock

### PR DESCRIPTION
**Pull request by eadm**
**YouTrack task**: [#APPS-1463](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1463)

**Description List**:
* Moved all Smart Lock logic to `SmartLockActivityBase`
* Create `BackToExitActivityWithSmartLockBase` same as `BackToExitActivityBase` but with Smart Lock support

### CODE REVIEW

File: app/src/classic/java/org.stepic.droid/ui/activities/LaunchActivity.kt [hunk 1]
```diff
@@ -45,27 +39,22 @@ import org.stepic.droid.ui.dialogs.LoadingProgressDialog
...
 import org.stepic.droid.util.ProgressHelper
 import org.stepic.droid.util.getMessageFor
-import timber.log.Timber
 import uk.co.chrisjenx.calligraphy.CalligraphyTypefaceSpan
 import uk.co.chrisjenx.calligraphy.TypefaceUtils
 import javax.inject.Inject
 
 
-class LaunchActivity : BackToExitActivityBase(), LoginView {
+class LaunchActivity : BackToExitActivityWithSmartLockBase(), LoginView {
```
Comment (KirillMakarov) [inline app/src/classic/java/org.stepic.droid/ui/activities/LaunchActivity.kt]: We should use Composition or Decoration for such functionality in future

File: app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    private var resolvingWasShown = false
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        resolvingWasShown = savedInstanceState?.getBoolean(RESOLVING_ACCOUNT_KEY) ?: false
+    }
+
+    protected var googleApiClient: GoogleApiClient? = null
+
+    protected fun initGoogleApiClient(withAuth: Boolean = false, autoManage: OnConnectionFailedListener = OnConnectionFailedListener {}) {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt]: can we use `null` as default value for `autoManage`? or is it not nullable?

File: app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        super.onSaveInstanceState(outState)
+    }
+
+    protected fun signOutFromGoogle() {
+        if (googleApiClient?.isConnected == true) {
+            Auth.GoogleSignInApi.signOut(googleApiClient)
+        }
+    }
+
+    protected open fun onCredentialRetrieved(credential: Credential) {}
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt]: let's make it abstract

File: app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    protected open fun onCredentialSaved() {}
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt]: let's make it abstract

File: app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt (hunk 3 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt]: If we make it abstract classes that don't need this functionality will have to have empty implementation of this method.

**Pull request merged**